token/client/src/token.rs

```diff
@@ -210,12 +221,29 @@ where
         &self.pubkey
     }
 
-    pub fn with_payer<S2: Signer>(&self, payer: S2) -> Token<T, S2> {
+    pub fn with_payer(&self, payer: Arc<dyn Signer>) -> Token<T> {
         Token {
             client: Arc::clone(&self.client),
             pubkey: self.pubkey,
             payer,
             program_id: self.program_id,
+            nonce_account: self.nonce_account,
+            nonce_authority: self.nonce_authority,
+            // XXX FIXME this is wrong but i dont understand the implications of clone on RwLock
+            // in fact i dont understand why we use write at all why not just Option<String> and clone
+            memo: Arc::new(RwLock::new(None)),
+        }
+    }
+
+    pub fn with_nonce(&self, nonce_account: &Pubkey, nonce_authority: &Pubkey) -> Token<T> {
+        Token {
+            client: Arc::clone(&self.client),
+            pubkey: self.pubkey,
+            payer: self.payer.clone(),
+            program_id: self.program_id,
+            // XXX FIXME as above. actually why cant these just take &mut self? seems silly
```

Comment:
These can't take a normal `&mut self` because you're in an async context, where the underlying `self` can be accessed in different async tasks.  It's goofy, but if you need mutability in a struct that's shared async, you need to wrap some part in `Arc<RwLock<T>>` -- either the whole struct, or the fields.  For the memo, we wrapped the field, but in this case, since it's set in the constructor, you shouldn't need to do that

Reply:
ah yea understand `&mut self` now. also understand why `Option<String>` isnt allowed, it works fine for a constructor but i didnt realize `take()` actually deleted the string off the object. this makes sense to me now, you can use the same mechanism for transfer memos

part of me wants to clone the memos in the other `with_` functions or name the functions something different or something but i guess i use this in one and only one place so it isnt really important